### PR TITLE
Enhance vault and strats.

### PR DESCRIPTION
**Pull request by skyinthesea-woo**
- BaseStrategy for vault, fee and pause functions
- Easier strategy write-up

### CODE REVIEW

File: contracts/earn/Vault.sol [hunk 1]
```diff
@@ -106,8 +102,8 @@ contract Vault is ERC20, Ownable, ReentrancyGuard {
     }
 
     function withdraw(uint256 shares) public nonReentrant {
-        require(shares > 0, 'Vault: shares_CAN_NOT_BE_ZERO');
-        require(shares <= balanceOf(msg.sender), 'Vault: shares exceed balance');
+        require(shares > 0, 'Vault: shares_ZERO');
+        require(shares <= balanceOf(msg.sender), 'Vault: shared_NOT_ENOUGH');
```
Comment (merlincdj) [inline contracts/earn/Vault.sol:106]: typo: `shares_NOT_ENOUGH`

File: contracts/earn/strategies/BaseStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    uint256 public constant FEE_MAX = 10000;
+    uint256 public performanceFee = 0; // 1 in 10000th: 1: 0.01%, 10: 0.1%, 100: 1%
+    uint256 public withdrawalFee = 0; // 1 in 10000th: 1: 0.01%, 10: 0.1%, 100: 1%
+    address public performanceTreasury;
+    address public withdrawalTreasury;
+
+    IWooAccessManager public accessManager;
+
+    constructor(address initVault, address initAccessManager) public {
+        require(initVault != address(0), 'StrategyCake: initVault_ZERO_ADDR');
```
Comment (merlincdj) [inline contracts/earn/strategies/BaseStrategy.sol:41]: typo: BaseStrategy: initVault_ZERO_ADDR

File: contracts/earn/strategies/BaseStrategy.sol [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+    function deposit() public virtual override;
+
+    function _giveAllowances() internal virtual;
+
+    function _removeAllowances() internal virtual;
+
+    function retireStrat() external virtual override;
+
+    function emergencyExit() external virtual override;
```
Comment (merlincdj) [inline contracts/earn/strategies/BaseStrategy.sol:99]: order should be:
external
public
internal

File: contracts/earn/strategies/PancakeSwap/StrategyCake.sol [hunk 4]
```diff
@@ -8,131 +8,74 @@ import '@openzeppelin/contracts/access/Ownable.sol';
...
     }
 
-    function _chargeFees() private {
-        uint256 fee = IERC20(want).balanceOf(address(this)).mul(strategistReward).div(REWARD_MAX);
-        if (fee > 0) {
-            TransferHelper.safeTransfer(want, treasury, fee);
-        }
-    }
-
-    /* ----- Admin Functions ----- */
```
Comment (merlincdj) [inline contracts/earn/strategies/PancakeSwap/StrategyCake.sol:134]: need this comment

File: contracts/earn/strategies/PancakeSwap/StrategyCake.sol [hunk 5]
```diff
@@ -8,131 +8,74 @@ import '@openzeppelin/contracts/access/Ownable.sol';
...
-        return IERC20(want).balanceOf(address(this));
-    }
-
     function balanceOfPool() public view override returns (uint256) {
         (uint256 amount, ) = IMasterChef(masterChef).userInfo(0, address(this));
-
         return amount;
     }
 
-    /* ----- Private Functions ----- */
```
Comment (merlincdj) [inline contracts/earn/strategies/PancakeSwap/StrategyCake.sol:116]: /* ----- Internal Functions ----- */

File: contracts/earn/strategies/PancakeSwap/StrategyLP.sol [hunk 6]
```diff
@@ -105,60 +80,45 @@ contract StrategyLP is Ownable, Pausable, IStrategy {
...
-        return IERC20(want).balanceOf(address(this));
-    }
-
     function balanceOfPool() public view override returns (uint256) {
         (uint256 amount, ) = IMasterChef(masterChef).userInfo(pid, address(this));
-
         return amount;
     }
 
     /* ----- Private Functions ----- */
```
Comment (merlincdj) [inline contracts/earn/strategies/PancakeSwap/StrategyLP.sol:119]: /* ----- Internal Functions ----- */

**Comment by skyinthesea-woo** [comment]
code review comments have been addressed.

**Pull request merged**